lms/djangoapps/badges/backends/badgr.py

```diff
@@ -0,0 +1,171 @@
+                u"Encountered an error when contacting the Badgr-Server. Request sent to %s with headers %s.\n"
+                u"and data values %s\n"
+                u"Response status was %s.\n%s",
+                repr(response.request.url), repr(response.request.headers),
+                repr(data),
+                response.status_code, response.content
+            )
+            raise
+
+    def _create_badge(self, badge_class):
+        """
+        Create the badge class on Badgr.
+        """
+        image = badge_class.image
+        # We don't want to bother validating the file any further than making sure we can detect its MIME type,
+        # for HTTP. The Badgr-Server should tell us if there's anything in particular wrong with it.
+        content_type, __ = mimetypes.guess_type(image.name)
+        if not content_type:
+            raise ValueError(
+                "Could not determine content-type of image! Make sure it is a properly named .png file."
+            )
+        files = {'image': (image.name, image, content_type)}
+        data = {
+            'name': badge_class.display_name,
+            'criteria': badge_class.criteria,
+            'slug': self._slugify(badge_class),
+            'description': badge_class.description,
+        }
+        result = requests.post(self._badge_create_url, headers=self._get_headers(), data=data, files=files)
+        self._log_if_raised(result, data)
+
+    def send_assertion_created_event(self, user, assertion):
+        """
+        Send an analytics event to record the creation of a badge assertion.
+        """
+        tracker.emit(
+            'edx.badge.assertion.created', {
+                'user_id': user.id,
+                'badge_slug': assertion.badge_class.slug,
+                'badge_name': assertion.badge_class.display_name,
+                'issuing_component': assertion.badge_class.issuing_component,
+                'course_id': unicode(assertion.badge_class.course_id),
+                'enrollment_mode': assertion.badge_class.mode,
+                'assertion_id': assertion.id,
+                'assertion_image_url': assertion.image_url,
+                'assertion_json_url': assertion.assertion_url,
+                'issuer': assertion.data.get('issuer'),
+            }
+        )
+
+    def _create_assertion(self, badge_class, user, evidence_url):
+        """
+        Register an assertion with the Badgr server for a particular user for a specific class.
+        """
+        data = {
+            'email': user.email,
+            'evidence': evidence_url,
+        }
+        response = requests.post(
+            self._assertion_url(self._slugify(badge_class)), headers=self._get_headers(), data=data
+        )
+        self._log_if_raised(response, data)
+        assertion, __ = BadgeAssertion.objects.get_or_create(user=user, badge_class=badge_class)
+        assertion.data = response.json()
+        assertion.backend = 'BadgrBackend'
```

Comment:
@nedbat Record keeping, in case we eventually support multiple badge issuing backends, or switch backends and need to contact the old one.